### ISSUE TITLE
FITSDiff numpy all close, zero values problem

### ISSUE THREAD

**Issue by SaOgaz** [astropy/astropy#7765]
Hi, I've run across kind of an interesting problem.  So [`FITSDiff`](http://docs.astropy.org/en/stable/io/fits/api/diff.html) is using [`numpy.allclose`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.allclose.html) whose equation is

`absolute(a - b) <= (atol + rtol * absolute(b))`

This equation, compared to the [`math.isclose`](https://docs.python.org/dev/library/math.html#math.isclose), which is

`abs(a-b) <= max(rel_tol * max(abs(a), abs(b)), abs_tol)`

has a problem when the b value is zero, as it makes the whole right side of the numpy equation zer., therefore returning a failing difference no matter what the `rtol` is set to. This is avoided in math by the max argument.  The ['numpy.isclose'](https://docs.scipy.org/doc/numpy/reference/generated/numpy.isclose.html) actually has a warning and discussion about this, and labels it as inappropriate to use for numbers much smaller than one. 

Could we replace the 'numpy.allclose' call with a `math.isclose` call? I realize this will probably take some extra logic to apply to the arrays etc.  This was the best solution @jamienoss and I could come up with after some brief discussion.

**Comment by pllim** [comment]
Here is the text from Numpy documentation posted above:

*Unlike the built-in math.isclose, the above equation is not symmetric in a and b – it assumes b is the reference value – so that isclose(a, b) might be different from isclose(b, a). Furthermore, the default value of atol is not zero, and is used to determine what small values should be considered close to zero. The default value is appropriate for expected values of order unity: if the expected values are significantly smaller than one, it can result in false positives. atol should be carefully selected for the use case at hand. A zero value for atol will result in False if either a or b is zero.*

Changing to `math.isclose` now is backward-incompatible. I would say the ship has sailed on this one. Maybe @embray would like to chime in? :pray:

**Comment by jamienoss** [comment]
The [numpy doc comment](https://docs.scipy.org/doc/numpy/reference/generated/numpy.isclose.html)
```
New in version 1.7.0.

For finite values, isclose uses the following equation to test whether two floating point values are equivalent.

absolute(a - b) <= (atol + rtol * absolute(b))
Unlike the built-in math.isclose, the above equation is not symmetric in a and b – it assumes b is the reference value – so that isclose(a, b) might be different from isclose(b, a). Furthermore, the default value of atol is not zero, and is used to determine what small values should be considered close to zero. The default value is appropriate for expected values of order unity: if the expected values are significantly smaller than one, it can result in false positives. atol should be carefully selected for the use case at hand. A zero value for atol will result in False if either a or b is zero.
```

**Comment by jamienoss** [comment]
Maybe the soln. is to allow the diff op to be passed in or as an abstract method.

**Comment by jamienoss** [comment]
The numpy doc even has the following warning:

_"Warning
The default atol is not appropriate for comparing numbers that are much smaller than one (see Notes)."_

**Comment by jamienoss** [comment]
Semantically, I think the best thing is to decide what _should_ be used and _then_ whether or how to make it possible.

My take, thus far, is that the use of ``numpy.isclose`` is incorrect here.

**Comment by jamienoss** [comment]
@pllim is _slightly_ right here
>Changing to math.isclose now is backward-incompatible.

If it gets changed it could cause other tests, out there in the world, that were/are close to the fence to fall the other side. That might be seen as breaking backward compatibilty, however, at the same time it is a bug under the semantics of ``fitsdiff`` in which case fixing bugs can change behaviour when that before was wrong.

**Comment by SaOgaz** [comment]
Uh, actually @pllim brought up a good point.... if I did my math right, even the `math` version will fail to give you the relative tolerance you want when one value is zero...

Say your `rtol` is set to 10^-7, your data is  10^-6 and 0...

you get
10^-7  <=  10^-7 * 10^-6

**Comment by jamienoss** [comment]
## math.isclose()
```
abs(a - b) <= max(r_tol * max(abs(a), abs(b)), a_tol)
```
This is symetric in ``a`` & ``b`` so let's make ``b`` zero for obvious comp with the ``numpy`` ver.
When ``b = a_tol = 0`` the above reduces to:
```
abs(a) < r_tol * abs(a)
```
which is what it should do, because ``a`` _is_ the difference. When ``b = 0`` the diff is ``a`` no matter what, as it should be. If ``b`` where only minutely off from zero, the diff would still be approx. ``a``.

 ## numpy.isclose()
```
abs(a - b) <= a_tol + r_tol * abs(b)
```
First point, this is _not_ symmetric in ``a`` & ``b`` (as the doc notes).
When ``b = a_tol = 0`` the above reduces to:
```
abs(a) <= 0
```

## 
_Also, they both assume ``r_tol := abs(r_tol)`` & ``a_tol := abs(a_tol)`` ("beacuse it won't work if they don't")_

**Comment by eteq** [comment]
I guess it *does* seem like a reasonable person would think that fitsdiff should be symmetric, since most everything else is. *but* @pllim is right that this would break backwards compatibility without an obvious reason.  

So what about implementing this as an *option* for `FITSDiff`? It would default to the current behavior, but if we really think the "right" behavior is the symmetric one (and the user community isn't up in arms about the change), we could have it first  deprecation warning for a version or two, and then eventually switch the default to the symmetric behavior.

The one catch is that we can't use `math.isclose` itself since it's not vectorized.  But the expression in https://github.com/astropy/astropy/issues/7765#issuecomment-415491786 is all vectorizable - just put ``np.`` in front of all the functions and I think it will work out of the box...

**Comment by jamienoss** [comment]
@eteq 
>The one catch is that we can't use math.isclose itself since it's not vectorized. But the expression for math.isclose in #7765 (comment) is all vectorizable - just put np. in front of all the functions and I think it will work out of the box...

Yeah, we'd already realized and thought of this one. By ``np.`` do you mean ``@np.vectorize`` or ...?

>we could have it first deprecation warning for a version or two, and then eventually switch the default to the symmetric behavior.

If it were anything else I wouldn't care, however, I wonder if we should think a little more about it considering that it acts as the corner-stone to most astro/fits regression/CI testing? I mean, it's a pretty serious issue when you get the test infra. struct. wrong. As much as the community may be up in arms about tests starting to fail once we make the change, might they not be as much "up in arms" that their tests may not be accurate? That being said, I haven't thought it through whether the issue creates only false negatives, which would be ok, i.e. ``a <= 0`` always fails.

> ...without an obvious reason.

Other than the above point 😉

**Comment by saimn** [comment]
There is also some discussion on Numpy side (I didn't read yet): https://github.com/numpy/numpy/issues/10161
We could add a new option with the symmetric version, and a warning when this option is not used, and change the default behavior in ~2 versions ?

**Comment by jamienoss** [comment]
@saimn cheers for the numpy link. A very interesting read.

It seemed to get quite "heated" IMHO - a few clashing heads.

It seems to cover 3 issues:

 * symmetry
 * mixing of ``atol``  &``rtol``
 * dangerous defaults for ``atol``, i.e. not default to zero

I kinda skipped over the mixing comments, I guessed that it has something to do with the use of ``max`` vs ``+``.

The only reasonable soln. I liked is [this one](https://github.com/numpy/numpy/issues/10161#issuecomment-349555152) - adding a ``symmetric`` flag to ``numpy.isclose()``.

**Comment by jamienoss** [comment]
BTW I think we should investigate the "mixing" concern, it may also be a worthy rationale for making the change inside of ``fitsdiff``.
